src/shared/settingsConfiguration.ts

```diff
@@ -5,231 +5,590 @@
-        if (val === undefined) {
-            const msg = `settings: getSetting(): setting "${key}": not found`
-            if (opt.silent !== undefined && opt.silent === 'no') {
-                throw Error(`AWS Toolkit: ${msg}`)
+
+        public get<K extends keyof Inner>(key: K & string, defaultValue?: Inner[K]) {
+            try {
+                const value = this.config.get(key, defaultValue)
+
+                if (!descriptor) {
+                    return value as Inner[K]
+                }
+
+                if (descriptor?.[key] === undefined) {
+                    throw new Error('No type descriptor found')
+                }
+
+                return cast<Inner[K]>(value, descriptor[key])
+            } catch (error) {
+                throw new Error(`Failed to parse setting: ${error}`)
             }
-            // Caller is expected to log in the rare case that a nonexistent
-            // setting is noteworthy.
-            return undefined
         }
 
-        const actualType = typeof val
-        if (actualType !== type) {
-            const msg = `settings: getSetting(): setting "${key}": got ${actualType}, expected ${type}`
-            if (opt.silent === undefined || opt.silent !== 'yes') {
-                throw Error(`AWS Toolkit: ${msg}`)
+        public async update<K extends keyof Inner>(key: K & string, value: Inner[K]) {
+            try {
+                await this.config.update(key, value)
+
+                return true
+            } catch (error) {
+                this.log(`failed to update field "${key}": ${error}`)
+
+                return false
             }
-            if (opt.logging === undefined || opt.logging) {
-                this.log.error(msg)
+        }
+
+        public async delete(key: keyof Inner & string) {
+            try {
+                await this.config.update(key, undefined)
+
+                return true
+            } catch (error) {
+                this.log(`failed to delete field "${key}": ${error}`)
+
+                return false
+            }
+        }
+
+        public async reset() {
+            try {
+                return await this.config.reset()
+            } catch (error) {
+                this.log(`failed to reset settings: ${error}`)
             }
-            return undefined
         }
-        return val
+
+        public dispose() {
+            return vscode.Disposable.from(...this.disposables).dispose()
+        }
     }
+}
+
+/**
+ * A stricter form of {@link vscode.Memento} with type-safe semantics.
+ */
+export interface TypedSettings<T extends Record<string, any>> extends Omit<Settings, 'get' | 'update'> {
+    /**
+     * Gets the value stored at `key`.
+     *
+     * This will always return the expected type, otherwise an error is thrown. Whether a
+     * missing key is considered valid or not is dependent on the type specified.
+     */
+    get<K extends keyof T>(key: K, defaultValue?: T[K]): T[K]
```

Comment:
Open question: should `defaultValue` be used in the scenario where user-input is not aligned with the expected type?

Personally, I'd much rather have execution stop if my configuration was bad rather than picking some arbitrary value. The current implementation only uses a default is no user config is defined.

Reply:
Currently thinking that the idea of a `defaultValue` doesn't make sense for this abstraction. It probably exists on `vscode.Memento` because the nullish-coalescing operator is fairly new. The goal here is a statically-typed store with async semantics. If I specify the following type:
```ts
interface Shape {
    foo: string
}
```
Then I would expect that indexing at `foo` _always_ is `string`. Not `string | undefined` and not `string | null`. Initializing does need to exist somewhere, but it does not need to leak out into the interfaces.

Reply:
Decided to use a default value as a fallback to keep any risk from these changes low. We already pass in a default anyway. Behavior similar to `silent: 'notfound'|'yes'|'no'` can be implemented through protected methods. Which I think is fine because it's not the common case. Errors always bubble up in CI of course.

Whether or not `defaultValue` should exist can be addressed in a different PR. Many of the concepts presented here apply to `vscode.Memento`. Might as well get some re-use.